### DISCUSSION TITLE
[@types/lodash] Lodash fp types ordering incorrect for LodashReduce (and likely others)

### DISCUSSION THREAD

**Discussion by fxfilmxf** [DefinitelyTyped/DefinitelyTyped#59909] [Issues with a @types package] [answered]
I'm attempting to reduce an object, but the type checker is incorrectly displaying an error that my object is not assignable to a list of objects. I _believe_ this is happening because the `object` type matches a function type with type `T` before matching with one with type `T extends object` since `T` matches type object.

These images should make things clearer (very simple example):

Type error:
<img width="1248" alt="Screen Shot 2022-04-14 at 3 17 50 PM" src="https://user-images.githubusercontent.com/544867/163485863-58c5a361-8947-4251-87da-cff293bd7990.png">

Currently generated types for fp reduce:
<img width="1398" alt="Screen Shot 2022-04-14 at 3 19 31 PM" src="https://user-images.githubusercontent.com/544867/163486294-3875ab6e-47a2-48a8-bb76-f467ad6cd697.png">

What I think the generated types should look like:
<img width="1333" alt="Screen Shot 2022-04-14 at 3 20 09 PM" src="https://user-images.githubusercontent.com/544867/163486630-d0b033d2-3aa3-46dd-9c1c-1dd5ab69e264.png">

No type error with change in ordering!
<img width="577" alt="Screen Shot 2022-04-14 at 3 20 14 PM" src="https://user-images.githubusercontent.com/544867/163486488-d8b28c13-4158-41c3-8faf-43190e8389c3.png">

[a comment by an automated account is not shown]

**Comment by aj-r** [comment, marked as answer]
The fp types are nearly impossible to get 100% correct due the way the params are split between multiple function calls. While reordering might fix this use case, I'm worried that it will break other use cases.

In cases like this, you should be able to explicitly specify some types to convince TypeScript to pick the right overload. I don't know what type your object really is, but it might look something like this:

```ts
reduce<Record<string, number>, number>((n: number) => 1, 0)({});
```

Also, for future reference, please paste the actual code here instead of using images. It makes it much easier to read and to copy/paste if needed.

**Reply by fxfilmxf** [reply]
Fair, I'd be worried about breaking other use cases too. Thanks for the response!
